Dockerfile

```diff
@@ -1,20 +1,32 @@
-FROM crystallang/crystal:1.2.2-alpine AS builder
+FROM docker.io/node:18-alpine AS node-builder
+RUN mkdir /build
+WORKDIR /build
+COPY --chown=0:0 package.json package-lock.json /build/
+RUN npm ci
+COPY --chown=0:0 webpack.config.js ./
+RUN mkdir src
+COPY --chown=0:0 src/assets src/assets
+RUN npm run build
+
+FROM docker.io/crystallang/crystal:1.7.2-alpine AS crystal-builder
 RUN apk update && apk upgrade && apk add sqlite-static
 WORKDIR /build/
 ARG version
-RUN git clone --branch ${version:-dist} --depth 1 https://github.com/toddsundsted/ktistec .
+COPY --chown=0:0 shard.yml shard.lock ./
 RUN shards install --production
-RUN crystal build src/ktistec/server.cr --static --no-debug --release
+COPY --chown=0:0 etc etc
+COPY --chown=0:0 src src
+RUN crystal build src/ktistec/server.cr --progress --static --no-debug --release
 
-FROM alpine:latest AS server
+FROM docker.io/library/alpine:latest AS server
 RUN apk --no-cache add tzdata
 WORKDIR /app
-COPY --from=builder /build/etc /app/etc
-COPY --from=builder /build/public /app/public
-COPY --from=builder /build/server /bin/server
-RUN mkdir /db
-RUN ln -s /app/public/uploads /uploads
-ENV KTISTEC_DB=/db/ktistec.db
-CMD ["/bin/server"]
-VOLUME /db /uploads
+COPY --chown=0:0 etc /app/etc
+COPY --chown=0:0 public /app/public
+COPY --from=node-builder /build/public/dist /app/public/dist
+COPY --from=crystal-builder /build/server /bin/server
+RUN rm -rf /app/public/uploads && ln -sf /data/uploads /app/public/uploads
+ENV KTISTEC_DB=/data/ktistec.db
+CMD ["/bin/sh", "-c", "mkdir -p /data/uploads && exec /bin/server"]
+VOLUME /data
 EXPOSE 3000
```

Comment:
In this part, we assemble the container that will be executed in production. It's based on the alpine image so it stays quite small. On line 26 we copy the build result from the node build, and 27 the result of the crystal build.

An assumption is made that all the data will live in /data, and that would be the volume that you would use to store both the DB and also the uploaded files. This used to be split in two, increasing the operational overhead.